presto-google-sheets/src/main/java/io/prestosql/plugin/google/sheets/GSheetsDataProvider.java

```diff
@@ -0,0 +1,205 @@
+    {
+        try {
+            List<List<Object>> data = readAllValuesFromSheetExpression(sheetsConfig.getMetadataSheetId());
+            // first line is assumed to be sheet header
+            if (data.size() > 1) {
+                for (int i = 1; i < data.size(); i++) {
+                    if (data.get(i).size() >= 2) {
+                        String tableId = String.valueOf(data.get(i).get(0));
+                        String sheetId = String.valueOf(data.get(i).get(1));
+                        if (tableName.equalsIgnoreCase(tableId)) {
+                            return Optional.of(sheetId);
+                        }
+                    }
+                }
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(SHEETS_METASTORE_ERROR, e);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<List<List<Object>>> getDataForSheetExpression(String sheetExpression)
+    {
+        try {
+            return Optional.of(readAllValuesFromSheetExpression(sheetExpression));
+        }
+        catch (Exception e) {
+            throw new PrestoException(SHEETS_UNKNOWN_TABLE_ERROR, e);
+        }
+    }
+
+    /**
+     * @return
+     * @throws IOException
+     */
+    private Credential getCredentials() throws IOException
+    {
+        GoogleCredential credential = null;
+        // Load client secrets.
+        try (InputStream in = new FileInputStream(sheetsConfig.getCredentialsFilePath())) {
+            credential = GoogleCredential.fromStream(in).createScoped(SCOPES);
+        }
+        catch (Exception e) {
+            throw new PrestoException(SHEETS_BAD_CREDENTIALS_ERROR, e);
+        }
+        return credential;
+    }
+
+    public Set<String> listTables()
+    {
+        Set<String> tables = new HashSet<>();
+        try {
+            List<List<Object>> tableMetadata = sheetDataCache.get(sheetsConfig.getMetadataSheetId()).get();
+            if (tableMetadata.size() > 1) { // we need both table name and sheet id
+                for (int i = 1; i < tableMetadata.size(); i++) {
+                    if (tableMetadata.get(i).size() > 0) {
+                        tables.add(String.valueOf(tableMetadata.get(i).get(0)));
+                    }
+                }
+            }
+        }
+        catch (Exception e) {
+            log.error(e, "Error retrieving table metadata mapping");
+        }
+        return tables;
+    }
+
+    @Override
+    public List<List<Object>> readAllValues(String tableName) throws Exception
+    {
+        String sheetId = tableSheetMappingCache.get(tableName).get();
+        if (null == sheetId) {
+            throw new PrestoException(SHEETS_UNKNOWN_TABLE_ERROR, "Data not found for table " + tableName);
+        }
+        return sheetDataCache.get(sheetId).get();
+    }
+
+    private List<List<Object>> readAllValuesFromSheetExpression(String sheetExpression) throws Exception
```

Comment:
Do you think that introducing some entity for `List<List<Object>>` be beneficial? Like `Rows` that store `Row`.

Reply:
Yes, I will add SheetRow / SheetRows once add better column type mapping, may be in next iteration if thats ok.